src/vehicle.cpp

```diff
@@ -950,92 +951,115 @@ static void RunVehicleDayProc()
-			continue;
+	Vehicle *front = v->First();
+
+	if (v->vcache.cached_cargo_age_period != 0) {
+		v->cargo_age_counter = std::min(v->cargo_age_counter, v->vcache.cached_cargo_age_period);
+		if (--v->cargo_age_counter == 0) {
+			v->cargo.AgeCargo();
+			v->cargo_age_counter = v->vcache.cached_cargo_age_period;
 		}
+	}
 
-		assert(Vehicle::Get(vehicle_index) == v);
+	/* Do not play any sound when crashed */
+	if (front->vehstatus & VS_CRASHED) return;
 
-		switch (v->type) {
-			default: break;
+	/* Do not play any sound when in depot or tunnel */
+	if (v->vehstatus & VS_HIDDEN) return;
 
-			case VEH_TRAIN:
-			case VEH_ROAD:
-			case VEH_AIRCRAFT:
-			case VEH_SHIP: {
-				Vehicle *front = v->First();
+	/* Do not play any sound when stopped */
+	if ((front->vehstatus & VS_STOPPED) && (front->type != VEH_TRAIN || front->cur_speed == 0)) return;
 
-				if (v->vcache.cached_cargo_age_period != 0) {
-					v->cargo_age_counter = std::min(v->cargo_age_counter, v->vcache.cached_cargo_age_period);
-					if (--v->cargo_age_counter == 0) {
-						v->cargo.AgeCargo();
-						v->cargo_age_counter = v->vcache.cached_cargo_age_period;
-					}
-				}
+	/* Check vehicle type specifics */
+	switch (v->type) {
+		case VEH_TRAIN:
+			if (Train::From(v)->IsWagon()) return;
+			break;
 
-				/* Do not play any sound when crashed */
-				if (front->vehstatus & VS_CRASHED) continue;
+		case VEH_ROAD:
+			if (!RoadVehicle::From(v)->IsFrontEngine()) return;
+			break;
 
-				/* Do not play any sound when in depot or tunnel */
-				if (v->vehstatus & VS_HIDDEN) continue;
+		case VEH_AIRCRAFT:
+			if (!Aircraft::From(v)->IsNormalAircraft()) return;
+			break;
 
-				/* Do not play any sound when stopped */
-				if ((front->vehstatus & VS_STOPPED) && (front->type != VEH_TRAIN || front->cur_speed == 0)) continue;
+		default:
+			break;
+	}
 
-				/* Check vehicle type specifics */
-				switch (v->type) {
-					case VEH_TRAIN:
-						if (Train::From(v)->IsWagon()) continue;
-						break;
+	v->motion_counter += front->cur_speed;
+	/* Play a running sound if the motion counter passes 256 (Do we not skip sounds?) */
+	if (GB(v->motion_counter, 0, 8) < front->cur_speed) PlayVehicleSound(v, VSE_RUNNING);
 
-					case VEH_ROAD:
-						if (!RoadVehicle::From(v)->IsFrontEngine()) continue;
-						break;
+	/* Play an alternating running sound every 16 ticks */
+	if (GB(v->tick_counter, 0, 4) == 0) {
+		/* Play running sound when speed > 0 and not braking */
+		bool running = (front->cur_speed > 0) && !(front->vehstatus & (VS_STOPPED | VS_TRAIN_SLOWING));
+		PlayVehicleSound(v, running ? VSE_RUNNING_16 : VSE_STOPPED_16);
+	}
+}
 
-					case VEH_AIRCRAFT:
-						if (!Aircraft::From(v)->IsNormalAircraft()) continue;
-						break;
+template <class T, typename Pred>
+static void IterateTickable(Pred pred)
+{
+	for (auto it = std::begin(T::tickable_vehicles), next_it = it; it != std::end(T::tickable_vehicles); it = next_it) {
+		++next_it;
+		pred(it->second);
+	}
+}
```

Comment:
This does not guarantee that the iterator is not invalidated.
When pred is something like CommonVehicleTick and Vehicle::Tick deletes the vehicle, any of the articulated/following parts via Vehicle::next will also be deleted, likely deleting the map item pointed to by next_it.

Reply:
Ah yes, multiple deletions...

Reply:
I guess deleting a vehicle could add it to a list of vehicles to be removed.

Reply:
Only iterating head vehicles would likely make handling deletions easier.